lib/valueflow.cpp

```diff
@@ -612,6 +612,8 @@ static void setTokenValue(Token* tok, ValueFlow::Value value, const Settings* se
             valueType.sign == ValueType::SIGNED && tok->valueType() &&
             ValueFlow::getSizeOf(*tok->valueType(), settings) >= ValueFlow::getSizeOf(valueType, settings))
             return;
+        if (tok->valueType() && tok->valueType()->isFloat() && valueType.isIntegral())
+            value.valueType = ValueFlow::Value::ValueType::FLOAT;
```

Comment:
Maybe this should be in `setTokenValueCast`?

Reply:
I think the type of `!=`'s value should not be INT, so that 
~~~
void f(double d) {
	if (d != 0) {
		int i = d;
		if (i == 0) {}
	}
}
~~~
is also handled correctly.

Reply:
The example above gives an oppositeInnerCondition error, which seems to be a different issue.

Reply:
Moving the code to `setTokenValueCast()` does not work, since `tok` is not available. I have now moved it to `setConditionalValues()`, seems to work so far.